canonical/checkbox
# Ask tester about available graphics ports and adapters instead of relying on resource job

## Description

The display resource job has always been unreliable to assess what graphics ports were available on a device, because the information provided by the graphics drivers and/or xrandr. Things have not gone better with 22.04, which switched to Wayland, forcing us to use gnome-randr, which introduced additional issues.

In addition, more and more devices must be tested using a USB-C to HDMI/DP video adapter. This is a different use case than the one handled by the existing `has_usb_type_c`.

After discussing with QA team, the proposed solution is two-fold:

1. introduce manifest entries for graphics ports. When running Checkbox, if tester selected any jobs that requires to test specific graphics ports (HDMI, DP, VGA, DVI), they will be presented with the following Manifest screen, with a new section ("Does this machine have the following graphics ports?")
2. split `has_usb_type_c` manifest entry into
    a. `has_usbc_data` for jobs testing data transfer (USB drives, Ethernet, keyboard, etc.)
    b. `has_usbc_video` for jobs testing video through USB-C connector

![image](https://user-images.githubusercontent.com/512471/209559771-5d0ed8fc-a258-487b-b1d7-64e9756c3491.png)

The docking jobs have also been updated to take these manifest entries into account.

Finally, the display resource job was removed. Some jobs that were using display resource but were not given a manifest entry (RCA and S-Video) have been kept with the requirement to `display` resource job removed. These jobs are not used in any test plans we actually use regularly, and the display requirement probably didn't work in the first place...

## Resolved issues

Fix #56 

## Tests

After setting up the testing environment as described in the [contributing guide](https://github.com/canonical/checkbox/blob/main/CONTRIBUTING.md), unit tests pass for `base` and `resource` providers:

```
checkbox/providers/base $ ./manage.py test
(...)
Ran 1079 tests in 41.949s

OK

checkbox/providers/resource $ ./manage.py test
(...)
Ran 64 tests in 4.699s

OK
```

The validation tests (`./manage.py validate`) also pass for both Base and Resource providers.

I also ran the following tests:

### Test 1

1. Run `checkbox-cli` and select 20.04 manual test plan
4. Select Monitor and Audio jobs
5. Make sure the proper manifest entries are available.
6. Select `Yes` for DisplayPort, and `No` for HDMI, VGA, DVI
7. Check that the DisplayPort are run and the HDMI/VGA/DVI jobs are skipped

### Test 2

1. Run `checkbox-cli` and select Dock Cert test plan
2. Select jobs that require graphics ports and/or adapters
3. Make sure the proper manifest entries are available.
4. Select `Yes` for DisplayPort, and `No` for HDMI, VGA, DVI
5. Check that the DisplayPort are run and the HDMI/VGA/DVI jobs are skipped

### Test 3

1. Run `checkbox-cli` and select 22.04 manual test plan
4. Select Monitor and Audio jobs
5. Make sure the proper manifest entries are available.
6. Select `Yes` for DisplayPort, and `No` for HDMI, VGA, DVI
7. Select `Yes` for "USB Type-C Video Adapter (USB-C to HDMI/DP/...)"
8. Check that the DisplayPort are run and the HDMI/VGA/DVI jobs are skipped
9. Check that USB-C to DisplayPort/HDMI/VGA are run

### Test 4

1. Run `checkbox-cli` and select 22.04 full test plan
2. Select USB jobs that previously required `has_usb_type_c`, such as `usb-c/c-to-a-adapter/hid`, `usb-c/hid`, `usb-c/insert`...
3. Make sure the proper manifest entries are available.
4. Select `Yes` to "USB Type-C Data (e.g. HID, Drives, Ethernet)"
5. Check that the selected jobs are executed like before

### Test 5

1. Run `checkbox-cli` and select 22.04 full test plan
2. Select USB jobs that previously required `has_usb_type_c`, such as `usb-c/c-to-a-adapter/hid`, `usb-c/hid`, `usb-c/insert`...
3. Make sure the proper manifest entries are available.
4. Select `No` to "USB Type-C Data (e.g. HID, Drives, Ethernet)"
5. Check that the selected jobs are skipped
